Nil pointer dereference in UpdateInflightIPs

### Version

<!-- helm ls -A --all -o json | jq '.[] | select(.name=="karpenter") | .app_version' -r -->
__Karpenter Version:__ v0.26.0
<!-- kubectl version | grep Server -->
__Kubernetes Version:__ v1.24.0


### Expected Behavior

I expected new nodes to be provisioned, although since this is my first time setting up Karpenter there may be unresolved configuration issues (I have not yet succeeded in getting it to provision anything). Failing that, I expected better error handling with a log message indicating where things fell short.

### Actual Behavior

A launch template has been created, but no EC2 instances have been launched. Karpenter is crashing at github.com/aws/karpenter/pkg/providers/subnet/subnet.go:146. Karpenter does appear to have discovered the subnets correctly.

### Steps to Reproduce the Problem

I don't have good repro steps, as I'm not doing anything specific to trigger this bug. I'm just trying to get Karpenter to work. 

I checked cloudtrail logs and did not see any denials.

Looking at `instance.go:241`...
```
	createFleetOutput, err := p.ec2Batcher.CreateFleet(ctx, createFleetInput)
	p.subnetProvider.UpdateInflightIPs(createFleetInput, createFleetOutput, instanceTypes, lo.Values(zonalSubnets), capacityType)
	if err != nil {
```

It looks like `err` needs to be checked before the call to `UpdateInflightIPs`

### Resource Specs and Logs

Karpenter pod logs:
```
2023-03-02T14:55:24.807Z        DEBUG   controller.awsnodetemplate      discovered subnets      {"commit": "39b48fe", "awsnodetemplate": "amd64ondemand", "subnets": ["subnet-3REDACTED7 (us-west-2c)", "subnet-2REDACTEDe (us-west-2a)", "subnet-aREDACTED6 (us-west-2b)"]}
2023-03-02T14:55:24.874Z        DEBUG   controller.aws  finished hydrating the launch template cache    {"commit": "39b48fe", "tag-key": "karpenter.k8s.aws/cluster", "tag-value": "REDACTED-eks-sandbox", "item-count": 1}
2023-03-02T14:55:25.200Z        DEBUG   controller.awsnodetemplate      discovered security groups      {"commit": "39b48fe", "awsnodetemplate": "amd64ondemand", "security-groups": ["sg-0REDACTED2"]}
2023-03-02T14:55:25.606Z        INFO    controller.provisioner  pod kube-system/coredns-c46857769-26627 has a preferred Anti-Affinity which can prevent consolidation   {"commit": "39b48fe"}
2023-03-02T14:55:25.606Z        INFO    controller.provisioner  pod kube-system/coredns-799c5565b4-v6tsl has a preferred Anti-Affinity which can prevent consolidation  {"commit": "39b48fe"}
2023-03-02T14:55:25.606Z        INFO    controller.provisioner  pod kube-system/aws-load-balancer-controller-7d5d495f54-fq8mz has a preferred Anti-Affinity which can prevent consolidation     {"commit": "39b48fe"}
2023-03-02T14:55:25.606Z        INFO    controller.provisioner  pod kube-system/aws-load-balancer-controller-7d5d495f54-vckd6 has a preferred Anti-Affinity which can prevent consolidation     {"commit": "39b48fe"}
2023-03-02T14:55:25.606Z        INFO    controller.provisioner  pod kube-system/coredns-c46857769-hc267 has a preferred Anti-Affinity which can prevent consolidation   {"commit": "39b48fe"}
2023-03-02T14:55:25.692Z        DEBUG   controller.deprovisioning       discovered EC2 instance types   {"commit": "39b48fe", "instance-type-count": 621}
2023-03-02T14:55:25.905Z        DEBUG   controller.provisioner  discovered EC2 instance types zonal offerings for subnets       {"commit": "39b48fe", "subnet-selector": "{\"subnet:zone\":\"private:nat\"}"}
2023-03-02T14:55:26.335Z        INFO    controller.provisioner  found provisionable pod(s)      {"commit": "39b48fe", "pods": 12}
2023-03-02T14:55:26.335Z        INFO    controller.provisioner  computed new node(s) to fit pod(s)      {"commit": "39b48fe", "nodes": 3, "pods": 12}
2023-03-02T14:55:26.335Z        INFO    controller.provisioner  launching machine with 4 pods requesting {"cpu":"375m","memory":"334Mi","pods":"10"} from types r6idn.2xlarge, r6a.2xlarge, c6i.8xlarge, c6i.4xlarge, c6id.8xlarge and 51 other(s)      {"commit": "39b48fe", "provisioner": "amd64ondemand"}
2023-03-02T14:55:26.335Z        INFO    controller.provisioner  launching machine with 4 pods requesting {"cpu":"575m","memory":"734Mi","pods":"10"} from types r6idn.2xlarge, r6a.2xlarge, c6i.8xlarge, c6i.4xlarge, c6id.8xlarge and 51 other(s)      {"commit": "39b48fe", "provisioner": "amd64ondemand"}
2023-03-02T14:55:26.339Z        INFO    controller.provisioner  launching machine with 4 pods requesting {"cpu":"375m","memory":"334Mi","pods":"10"} from types r6idn.2xlarge, r6a.2xlarge, c6i.8xlarge, c6i.4xlarge, c6id.8xlarge and 51 other(s)      {"commit": "39b48fe", "provisioner": "amd64ondemand"}
2023-03-02T14:55:26.386Z        DEBUG   controller.provisioner.cloudprovider    discovered kubernetes version   {"commit": "39b48fe", "provisioner": "amd64ondemand", "kubernetes-version": "1.24"}
2023-03-02T14:55:26.985Z        DEBUG   controller.provisioner.cloudprovider    discovered images       {"commit": "39b48fe", "provisioner": "amd64ondemand", "ami-ids": ["ami-02REDACTED30", "ami-02REDACTED70", "ami-0dREDACTED54"]}
E0302 14:55:27.268495       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1146 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1dcef60?, 0x3bc3ec0})
        k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:75 +0x7c
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xffff6824f548?})
        k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:49 +0x78
panic({0x1dcef60, 0x3bc3ec0})
        runtime/panic.go:884 +0x1f4
github.com/aws/karpenter/pkg/providers/subnet.(*Provider).UpdateInflightIPs(0x4000f2c980, 0x400651e5a0, 0x0, {0x40014c9800, 0x38, 0x40}, {0x4005eefe48?, 0x3, 0x9?}, {0x22232ec, ...})
        github.com/aws/karpenter/pkg/providers/subnet/subnet.go:146 +0x380
github.com/aws/karpenter/pkg/cloudprovider.(*InstanceProvider).launchInstance(0x4000f154f0, {0x27363e8, 0x4007efe240}, 0x4006c185a0, 0x0?, {0x40014c9800, 0x38, 0x40})
        github.com/aws/karpenter/pkg/cloudprovider/instance.go:241 +0x988
github.com/aws/karpenter/pkg/cloudprovider.(*InstanceProvider).Create(0x4000f154f0, {0x27363e8, 0x4007efe240}, 0x400198b1e0?, 0x400198b1e0, {0x40014c9600?, 0x4002abb560?, 0x1773758?})
        github.com/aws/karpenter/pkg/cloudprovider/instance.go:90 +0xc8
github.com/aws/karpenter/pkg/cloudprovider.(*CloudProvider).Create(0x4000f78fc0, {0x27363e8, 0x4007efe240}, 0x400198b1e0)
        github.com/aws/karpenter/pkg/cloudprovider/cloudprovider.go:105 +0x1e4
github.com/aws/karpenter-core/pkg/cloudprovider/metrics.(*decorator).Create(0x4000f1d950, {0x27363e8, 0x4007efe240}, 0x3c26950?)
        github.com/aws/karpenter-core@v0.25.1-0.20230224225813-3e5ec9b8eb1a/pkg/cloudprovider/metrics/cloudprovider.go:71 +0x174
github.com/aws/karpenter-core/pkg/controllers/provisioning.(*Provisioner).Launch(0x4000f4c6c0, {0x27363e8, 0x4007efe060}, 0x40016af980, {0x4003b36b40, 0x2, 0x2})
        github.com/aws/karpenter-core@v0.25.1-0.20230224225813-3e5ec9b8eb1a/pkg/controllers/provisioning/provisioner.go:319 +0x260
github.com/aws/karpenter-core/pkg/controllers/provisioning.(*Provisioner).LaunchMachines.func1(0x1)
        github.com/aws/karpenter-core@v0.25.1-0.20230224225813-3e5ec9b8eb1a/pkg/controllers/provisioning/provisioner.go:144 +0x218
k8s.io/client-go/util/workqueue.ParallelizeUntil.func1()
        k8s.io/client-go@v0.25.4/util/workqueue/parallelizer.go:90 +0x104
created by k8s.io/client-go/util/workqueue.ParallelizeUntil
        k8s.io/client-go@v0.25.4/util/workqueue/parallelizer.go:76 +0x1a0
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x16d3050]
```

`kubectl describe provisioner amd64ondemand` (trimmed):

```
Spec:
  Consolidation:
    Enabled:  true
  Provider Ref:
    Name:  amd64ondemand
  Requirements:
    Key:       karpenter.k8s.aws/instance-category
    Operator:  In
    Values:
      c
      m
      r
    Key:       karpenter.k8s.aws/instance-generation
    Operator:  Gt
    Values:
      5
    Key:       karpenter.k8s.aws/instance-cpu
    Operator:  In
    Values:
      4
      8
      16
      32
    Key:       karpenter.k8s.aws/instance-hypervisor
    Operator:  In
    Values:
      nitro
    Key:       topology.kubernetes.io/zone
    Operator:  In
    Values:
      us-west-2a
      us-west-2b
      us-west-2c
    Key:       kubernetes.io/arch
    Operator:  In
    Values:
      amd64
    Key:       karpenter.sh/capacity-type
    Operator:  In
    Values:
      on-demand
    Key:       kubernetes.io/os
    Operator:  In
    Values:
      linux
  Weight:  1
Events:    <none>
```

`kubectl describe awsnodetemplate amd64ondemand` (trimmed):

```
Spec:
  Ami Family:  Custom
  Ami Selector:
    aws::name:        REDACTED_Amzn2_EKS_1.24_*
    aws::owners:      7REDFACTED2
    In - Production:  true
  Security Group Selector:
    Name:  REDACTED-eks-workers
  Subnet Selector:
    subnet:zone:  private:nat
  Tags:
    Application:          kubernetes
    Application - Group:  REDACTED
    Environment:          sandbox
    Provisioner:          karpenter
  User Data:              REDACTED

Status:
  Security Groups:
    Id:  sg-0REDACTED2
  Subnets:
    Id:    subnet-aREDACTED6
    Zone:  us-west-2b
    Id:    subnet-3REDACTED7
    Zone:  us-west-2c
    Id:    subnet-2REDACTEDe
    Zone:  us-west-2a
Events:    <none>
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

--- a/pkg/cloudprovider/suite_test.go
+++ b/pkg/cloudprovider/suite_test.go
@@ -16,6 +16,7 @@ package cloudprovider
 
 import (
 	"context"
+	"fmt"
 	"net"
 	"testing"
 	"time"
@@ -569,6 +570,17 @@ var _ = Describe("Allocation", func() {
 			ExpectScheduled(ctx, env.Client, pod3)
 			createFleetInput = fakeEC2API.CreateFleetBehavior.CalledWithInput.Pop()
 			Expect(fake.SubnetsFromFleetRequest(createFleetInput)).To(ConsistOf("test-subnet-1"))
+		})
+		It("should update in-flight IPs when a CreateFleet error occurs", func() {
+			fakeEC2API.DescribeSubnetsOutput.Set(&ec2.DescribeSubnetsOutput{Subnets: []*ec2.Subnet{
+				{SubnetId: aws.String("test-subnet-1"), AvailabilityZone: aws.String("test-zone-1a"), AvailableIpAddressCount: aws.Int64(10),
+					Tags: []*ec2.Tag{{Key: aws.String("Name"), Value: aws.String("test-subnet-1")}}},
+			}})
+			pod1 := coretest.UnschedulablePod(coretest.PodOptions{NodeSelector: map[string]string{v1.LabelTopologyZone: "test-zone-1a"}})
+			ExpectApplied(ctx, env.Client, provisioner, nodeTemplate, pod1)
+			fakeEC2API.CreateFleetBehavior.Error.Set(fmt.Errorf("CreateFleet synthetic error"))
+			bindings := ExpectProvisioned(ctx, env.Client, cluster, prov, pod1)
+			Expect(len(bindings)).To(Equal(0))
 		})
 		It("should launch instances into subnets that are excluded by another provisioner", func() {
 			fakeEC2API.DescribeSubnetsOutput.Set(&ec2.DescribeSubnetsOutput{Subnets: []*ec2.Subnet{

--- a/pkg/providers/subnet/subnet.go
+++ b/pkg/providers/subnet/subnet.go
@@ -143,12 +143,15 @@ func (p *Provider) UpdateInflightIPs(createFleetInput *ec2.CreateFleetInput, cre
 	})))
 
 	// Process the CreateFleetOutput to pull out all the fulfilled subnetIDs
-	fleetOutputSubnets := lo.Compact(lo.Uniq(lo.Map(createFleetOutput.Instances, func(fleetInstance *ec2.CreateFleetInstance, _ int) string {
-		if fleetInstance == nil || fleetInstance.LaunchTemplateAndOverrides == nil || fleetInstance.LaunchTemplateAndOverrides.Overrides == nil {
-			return ""
-		}
-		return lo.FromPtr(fleetInstance.LaunchTemplateAndOverrides.Overrides.SubnetId)
-	})))
+	var fleetOutputSubnets []string
+	if createFleetOutput != nil {
+		fleetOutputSubnets = lo.Compact(lo.Uniq(lo.Map(createFleetOutput.Instances, func(fleetInstance *ec2.CreateFleetInstance, _ int) string {
+			if fleetInstance == nil || fleetInstance.LaunchTemplateAndOverrides == nil || fleetInstance.LaunchTemplateAndOverrides.Overrides == nil {
+				return ""
+			}
+			return lo.FromPtr(fleetInstance.LaunchTemplateAndOverrides.Overrides.SubnetId)
+		})))
+	}
 
 	// Find the subnets that were included in the input but not chosen by Fleet, so we need to add the inflight IPs back to them
 	subnetIDsToAddBackIPs, _ := lo.Difference(fleetInputSubnets, fleetOutputSubnets)

